Cargo.toml

```diff
@@ -58,6 +58,7 @@ default = [
   "bevy_sprite",
   "bevy_text",
   "bevy_ui",
+  "bevy_dev_tools",
```

Comment:
Do we really want this to be enabled by default 🤔 ?   If we go with "disable this in the release build" approach, wouldn't that hinder disabling this feature?
We would have to do something like that in the TOML:
```toml
[dependencies]
bevy = {version = "0.12.1", default-features = false, features = ["all features", "except the debug feature"]}
[dev-dependencies]
bevy = {version = "0.12.1"}
```

Reply:
There are two different approaches that I could see being used.

### Disable on release mode

You could wrap `fn build()` in a `#[cfg(debug_assertions)]` so that the plugin only does things in debug mode.

```rust
#[cfg(debug_assertions)]
fn build(&self, app: &mut App) {
    ...
}
```

I don't think this is a good idea for a multitude of reasons. First, it's inconsistent with other developer crates Bevy provides. [Dynamic linking](https://bevyengine.org/learn/book/getting-started/setup/#enable-fast-compiles-optional) is opt-in, so I believe that devtools should be treated the same. Additionally, [the setup guide](https://bevyengine.org/learn/book/getting-started/setup/#compile-with-performance-optimizations) recommends turning on `opt-level = 3` for all dependencies, including Bevy. I'm not positive, but that may turn off debug assertions and thus devtools.

### Non-default feature

I think this is the better option. This only issue is that developers will need to know that the devtools exist in order to use them. This is fine in my opinion, as it will work similar to `bevy_dylib`. And since developers will have to manually enable it, they will know to disable it in release mode. (At least I hope.)

Reply:
Non-default feature sounds best to me. Maybe `DevToolsPlugin` should emit a warning on build in production.

Reply:
> Emit a warning on production

A warning like "you are running the dev-tools on release, ignore if this is intentional"?

Reply:
Exactly. I'm sure I will totally forget to disable it otherwise.

Reply:
Yeah, I think we should leave this in default features to make the examples and beginner experience run smoothly, but emit an `info!()` log about "you're running dev tools! This may give players undesired access to your game! Maybe turn it off when shipping?"

Leaving it off by default only helps stop the absolute beginners from shipping it, but experienced teams will do this too.

Reply:
I can understand this point, offering the beginners the tools by default is good, but IMO the only problem is the pain of disabling it by listing all the default features (you will have to turn off the default-features, and them list all the default-features, except for the bevy_dev_tools). That wouldn't be a problem if cargo supported disabling one of default features with some sintax though. I do agree however that this "problem" may not be a problem, just a boilerplate in the Cargo.toml 
An example of the result, using all the default-features:
```toml
[dev-dependencies]
bevy = {version = "0.12.1"}

[dependencies]
bevy = {version = "0.12.1", features = [
  "animation",
  "bevy_asset",
  "bevy_audio",
  "bevy_gilrs",
  "bevy_scene",
  "bevy_winit",
  "bevy_core_pipeline",
  "bevy_pbr",
  "bevy_gltf",
  "bevy_render",
  "bevy_sprite",
  "bevy_text",
  "bevy_ui",
  "multi-threaded",
  "png",
  "hdr",
  "vorbis",
  "x11",
  "bevy_gizmos",
  "android_shared_stdcxx",
  "tonemapping_luts",
  "default_font",
  "webgl2",
  # all the actual default features, except for this one
]}
```
Well, this is a boilerplate, but nothing will go wrong for sure
One another option for beginners is strongly recomending the feature somewhere in the book, but that would probably not be as effective as that feature being a default-feature though

Reply:
what about using `info!("dev tools are enabled")` in dev mode, and `warn!("you're running dev tools in release mode, this is not recommended")` in release mode - text used as an example

Reply:
> I can understand this point, offering the beginners the tools by default is good, but IMO the only problem is the pain of disabling it by listing all the default features (you will have to turn off the default-features, and them list all the default-features, except for the bevy_dev_tools). That wouldn't be a problem if cargo supported disabling one of default features with some sintax though. I do agree however that this "problem" may not be a problem, just a boilerplate in the Cargo.toml 
> An example of the result, using all the default-features:
> ```toml
> [dev-dependencies]
> bevy = {version = "0.12.1"}
> 
> [dependencies]
> bevy = {version = "0.12.1", default-features = false, features = [
>   "animation",
>   "bevy_asset",
>   "bevy_audio",
>   "bevy_gilrs",
>   "bevy_scene",
>   "bevy_winit",
>   "bevy_core_pipeline",
>   "bevy_pbr",
>   "bevy_gltf",
>   "bevy_render",
>   "bevy_sprite",
>   "bevy_text",
>   "bevy_ui",
>   "multi-threaded",
>   "png",
>   "hdr",
>   "vorbis",
>   "x11",
>   "bevy_gizmos",
>   "android_shared_stdcxx",
>   "tonemapping_luts",
>   "default_font",
>   "webgl2",
>   # all the actual default features, except for this one
> ]}
> ```
> Well, this is a boilerplate, but nothing will go wrong for sure
> One another option for beginners is strongly recomending the feature somewhere in the book, but that would probably not be as effective as that feature being a default-feature though

You could also just disable DevToolsPlugin from DefaultPlugins, should work the same from what I know

Reply:
I also think this should not be part of the default features

Reply:
> Maybe DevToolsPlugin should emit a warning on build in production.

What's wrong with having dev tools in release mode?

Reply:
Dev tools will probably add bloat and slow things down: profiling, version-control/undo-systems, ect add a lot of extra stuff. Most users won't want to ship them, and release mode is the best indicator that "ok, this build is maybe going to ship". I'm not suggesting that they shouldn't work in release, just that they log a warning to help prevent accidents.

But I guess that's a bridge we can cross when we come to it. The crate is empty. If we need to take this off default at a later point we can.

Reply:
If you want an option to *not* compile it in `release`, it should be behind a new feature flag. This is very similar to `debug-render-3d` in rapier, which is a debug feature behind a flag.

Keep in mind that people (like me) would most likely want to include it in `release` for better debuggability and modding support in production (unless there's good reason not to, e.g. pulling in tracing libraries).

Emitting a warning in production is just wrong. The code is already compiled in by that point, so what is that warning trying to achieve besides being annoying in false positive cases?

Reply:
This should not be a default crate for a number of reasons:
1. We currently operate under the assumption that `cargo build --release` should be a "shippable" binary. Practically, this is how people will use Bevy. Removing the default features and re-adding what you need is a nasty, constantly shifting, and arcane workflow in my opinion. This isn't something we can thrust on users by default. We can reconsider once we have an editor capable of facilitating these complexities, but even then I am dubious.
2. This is a "grab bag" crate in its current form (very loose bounds on scope, size and contents will likely balloon indefinitely), with little to no intended value for released apps. Something like that shouldn't be shipped by default imo.

I think this should be an opt-in cargo. Any examples that rely on the dev tools should add the feature as a requirement. For things that _should_ be enabled by default, they can just have their own owns with a clearer scope.